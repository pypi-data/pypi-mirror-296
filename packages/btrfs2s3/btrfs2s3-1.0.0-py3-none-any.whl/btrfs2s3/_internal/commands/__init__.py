"""Code related to CLI commands for btrfs2s3."""
