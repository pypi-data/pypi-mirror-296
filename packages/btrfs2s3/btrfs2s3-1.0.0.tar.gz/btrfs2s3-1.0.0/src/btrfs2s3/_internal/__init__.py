"""The implementation for the btrfs2s3 command line interface."""
