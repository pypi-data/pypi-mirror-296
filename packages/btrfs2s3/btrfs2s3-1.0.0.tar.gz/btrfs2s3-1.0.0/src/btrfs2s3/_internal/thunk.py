"""A Thunk wraps a value that might be lazily evaluated."""

from __future__ import annotations

import enum
from functools import total_ordering
from typing import Callable
from typing import Generic
from typing import Literal
from typing import overload
from typing import Protocol
from typing import runtime_checkable
from typing import Union

from typing_extensions import TypeAlias
from typing_extensions import TypeVar


class Error(Exception):
    """The top-level class for errors produced by this module."""


class StillTBDError(Error):
    """Thunk.check() was called, but the thunk hasn't been evaluated yet."""


_T = TypeVar("_T")
_T_contra = TypeVar("_T_contra", contravariant=True)


@runtime_checkable
class _SupportsLT(Protocol[_T_contra]):
    def __lt__(self, other: _T_contra) -> bool: ...


class _TbdType(enum.Enum):
    TBD = enum.auto()

    def __str__(self) -> str:
        return "TBD"


TBD = _TbdType.TBD
"""A sentinel value, meaning the real value hasn't been determined yet."""


@total_ordering
class Thunk(Generic[_T]):
    """A Thunk is a wrapper for a value which may or may not be lazily computed.

    A pre-computed Thunk will return its pre-computed value from peek(),
    check() and __call__().

    A lazily-evaluated Thunk will return the TBD sentinel value from peek(),
    and will raise StillTBDError from check(). __call__() will run the lazy
    evaluation function, and save the value. After this the Thunk will behave
    exactly like a pre-computed Thunk. The evaluation function will not be run
    again.

    This is a very rudimentary lazy evaluation wrapper class. Its only goal is
    to facilitate creating a "dry-run plan" for code that does destructive
    changes, and to present that to the user.

    The intended use of Thunk is as a building block to create complex "intent"
    objects, e.g. an intent to create a backup of some data. The various
    parameters (data source, storage location, etc) of this intent can be
    stored as Thunks which may be precomputed
    or lazily evaluated. This intent can be previewed to a user, and optionally
    executed. Using Thunks allows the preview and execution code to be agnostic
    of data dependencies and edge cases. The preview code can just display all
    the precomputed values in the intent object, and the execution code can
    just evaluate the thunks as needed. Note that this means all the complexity
    is pushed to the code that creates the intent objects.
    """

    _value: _T | Literal[_TbdType.TBD]
    _get_value: Callable[[], _T]

    # Without these overloads, mypy can't infer the type of Thunk(other_thunk),
    # presumably it can't figure out if this should match __init__(_T) or
    # __init__(Thunk[_T]).
    @overload
    def __init__(self, other: Thunk[_T]) -> None: ...
    @overload
    def __init__(self, other: Callable[[], _T]) -> None: ...
    @overload
    def __init__(self, other: _T) -> None: ...
    def __init__(self, other: Thunk[_T] | Callable[[], _T] | _T) -> None:
        """Create a thunk.

        Thunk(other_thunk) creates a copy of other_thunk. The result will have
        the same precomputed value and/or evaluation function as the argument
        Thunk.

        Thunk(callable) creates a lazily-evaluated Thunk. The argument must be
        a zero-args callable.

        Thunk(any-other-value) creates a pre-computed Thunk.

        Args:
            other: A zero-args callable, or another Thunk, or any other
                precomputed value.
        """
        if isinstance(other, Thunk):
            self._get_value = other._get_value  # noqa: SLF001
            self._value = other._value  # noqa: SLF001
        elif callable(other):
            self._get_value = other
            self._value = TBD
        else:
            self._get_value = lambda: other  # pragma: no cover
            self._value = other

    def peek(self) -> _T | Literal[_TbdType.TBD]:
        """Returns the precomputed value, or the TBD sentinel value."""
        return self._value

    def check(self) -> _T:
        """Returns the precomputed value, or raises StillTBDError.

        This asserts that the Thunk is pre-computed. It is mostly useful in
        tests.

        Raises:
            StillTBDError: If the thunk is lazy, and hasn't been evaluated yet.
        """
        value = self._value
        if value is TBD:
            raise StillTBDError
        return value

    def is_tbd(self) -> bool:
        """Returns True if the Thunk is lazy and hasn't been evaluated yet."""
        return self._value is TBD

    def __call__(self) -> _T:
        """Returns the value of the Thunk, possibly running its lazy computation.

        Returns:
            The value of the Thunk.
        """
        if self._value is TBD:
            self._value = self._get_value()
        return self._value

    def __repr__(self) -> str:
        """Return a string representation of the Thunk."""
        value = self._value
        if value is TBD:
            return "Thunk(TBD)"
        return f"Thunk({value!r})"

    def __lt__(self, other: Thunk[_T]) -> bool:
        """Returns whether this Thunk should appear before other Thunks.

        The natural ordering of Thunks preserves the natural ordering of their
        values, except that lazy Thunks appear after precomputed Thunks.
        """
        value = self._value
        othervalue = other._value
        if value is not TBD:
            if othervalue is not TBD:
                if not isinstance(value, _SupportsLT):
                    return NotImplemented
                return value.__lt__(othervalue)
            return True
        if othervalue is not TBD:
            return False
        return id(self) < id(other)


ThunkArg: TypeAlias = Union[Thunk[_T], _T, Callable[[], _T]]
"""An alias to allowed types for the argument to Thunk()."""
