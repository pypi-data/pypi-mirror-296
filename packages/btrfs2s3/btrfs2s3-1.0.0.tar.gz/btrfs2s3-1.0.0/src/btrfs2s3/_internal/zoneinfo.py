"""Compatibility module for zoneinfo."""

import sys

if sys.version_info >= (3, 9):  # pragma: >=3.9 cover
    from zoneinfo import ZoneInfo
else:  # pragma: <3.9 cover
    from backports.zoneinfo import ZoneInfo


# This function is just syntactic sugar to call the ZoneInfo constructor with
# appropriate lint markers
def get_zoneinfo(key: str) -> ZoneInfo:
    """Return a ZoneInfo object from a string timezone name."""
    if sys.version_info >= (3, 9):  # pragma: >=3.9 cover
        return ZoneInfo(key)
    else:  # noqa: RET505 # pragma: <3.9 cover
        # https://github.com/pganssle/zoneinfo/issues/125
        # Unfortunately the workaround in the linked issue does not work with
        # mypy --strict (I get: Call to untyped function "ZoneInfo" in typed
        # context)

        # Also mypy requires if-return-else to understand the version
        # condition, which thus requires the RET505 block. Yakkity yak
        return ZoneInfo(key)  # type: ignore[abstract]
