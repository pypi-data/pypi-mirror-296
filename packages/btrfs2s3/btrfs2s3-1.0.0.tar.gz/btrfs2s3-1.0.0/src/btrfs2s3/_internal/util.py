from __future__ import annotations

from enum import IntFlag

from btrfsutil import SubvolumeInfo

from btrfs2s3._internal.backups import BackupInfo

NULL_UUID = b"\0" * 16


def mksubvol(
    *,
    id: int = 0,  # noqa: A002
    parent_id: int = 0,
    dir_id: int = 0,
    flags: int = 0,
    uuid: bytes = NULL_UUID,
    parent_uuid: bytes = NULL_UUID,
    received_uuid: bytes = NULL_UUID,
    generation: int = 0,
    ctransid: int = 0,
    otransid: int = 0,
    stransid: int = 0,
    rtransid: int = 0,
    ctime: float = 0.0,
    otime: float = 0.0,
    stime: float = 0.0,
    rtime: float = 0.0,
) -> SubvolumeInfo:
    return SubvolumeInfo(
        (
            id,
            parent_id,
            dir_id,
            flags,
            uuid,
            parent_uuid,
            received_uuid,
            generation,
            ctransid,
            otransid,
            stransid,
            rtransid,
            ctime,
            otime,
        ),
        {"stime": stime, "rtime": rtime},
    )


def backup_of_snapshot(
    snapshot: SubvolumeInfo, send_parent: SubvolumeInfo | None = None
) -> BackupInfo:
    return BackupInfo(
        uuid=snapshot.uuid,
        parent_uuid=snapshot.parent_uuid,
        send_parent_uuid=None if send_parent is None else send_parent.uuid,
        ctransid=snapshot.ctransid,
        ctime=snapshot.ctime,
    )


class SubvolumeFlags(IntFlag):
    ReadOnly = 1
    Proposed = 2**30
