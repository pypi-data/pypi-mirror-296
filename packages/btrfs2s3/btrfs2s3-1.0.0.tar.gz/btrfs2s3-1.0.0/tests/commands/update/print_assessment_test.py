from __future__ import annotations

from datetime import timezone
from functools import partial
from pathlib import Path
from typing import TYPE_CHECKING

import arrow
from btrfs2s3._internal.assessor import Assessment
from btrfs2s3._internal.assessor import BackupAssessment
from btrfs2s3._internal.assessor import SnapshotAssessment
from btrfs2s3._internal.assessor import SourceAssessment
from btrfs2s3._internal.commands.update import print_assessment
from btrfs2s3._internal.resolver import Flags
from btrfs2s3._internal.resolver import KeepMeta
from btrfs2s3._internal.resolver import Reasons
from btrfs2s3._internal.thunk import Thunk
from btrfs2s3._internal.util import backup_of_snapshot
from btrfs2s3._internal.util import mksubvol
from btrfs2s3._internal.util import SubvolumeFlags
import pytest

if TYPE_CHECKING:
    from rich.console import Console


print_assessment_under_test = partial(
    print_assessment,
    tzinfo=timezone.utc,
    snapshot_dir=Path("/snapshots"),
    bucket="my-bucket",
)


@pytest.fixture()
def golden_asmt_with_no_changes() -> Assessment:
    asmt = Assessment()
    source1 = mksubvol(
        id=12345,
        uuid=b"$\x96\xea\xb9\x1f\xbcD\x7f\xba\x98\xf1\xc5b\xaf\xc6A",
        ctransid=123,
        ctime=arrow.get("2006-01-02").timestamp(),
    )
    source1_asmt = SourceAssessment(path=Path("/path/to/source1"), info=source1)
    asmt.sources[source1.uuid] = source1_asmt
    snap1_1 = mksubvol(
        id=source1.id + 1,
        uuid=b"4k\xd4D\xb1\x87O\xdb\x88\xdd=\xa6\xfd\xef\x84B",
        parent_uuid=source1.uuid,
        flags=SubvolumeFlags.ReadOnly,
        ctransid=source1.ctransid - 3,
        ctime=source1.ctime - 86400,
    )
    source1_asmt.snapshots[snap1_1.uuid] = SnapshotAssessment(
        initial_path=Path("/snapshots/snap1_1"),
        info=snap1_1,
        target_path=Thunk(Path("/snapshots/snap1_1")),
        real_info=Thunk(snap1_1),
        keep_meta=KeepMeta(
            reasons=Reasons.Preserved,
            time_spans={(arrow.get("2006").timestamp(), arrow.get("2007").timestamp())},
        ),
    )
    snap1_2 = mksubvol(
        id=snap1_1.id + 1,
        uuid=b"\x11[\xab\xa0\xec\xedA\xa6\xbbm3D\x9c\x92\xbf=",
        parent_uuid=source1.uuid,
        flags=SubvolumeFlags.ReadOnly,
        ctransid=source1.ctransid - 1,
        ctime=source1.ctime,
    )
    source1_asmt.snapshots[snap1_2.uuid] = SnapshotAssessment(
        initial_path=Path("/snapshots/snap1_2"),
        info=snap1_2,
        target_path=Thunk(Path("/snapshots/snap1_2")),
        real_info=Thunk(snap1_2),
        keep_meta=source1_asmt.snapshots[snap1_1.uuid].keep_meta,
    )
    backup1_1 = backup_of_snapshot(snap1_1, send_parent=None)
    source1_asmt.backups[backup1_1.uuid] = BackupAssessment(
        backup=Thunk(backup1_1),
        key=Thunk("backup1_1"),
        keep_meta=source1_asmt.snapshots[snap1_1.uuid].keep_meta,
    )
    backup1_2 = backup_of_snapshot(snap1_2, send_parent=snap1_1)
    source1_asmt.backups[backup1_2.uuid] = BackupAssessment(
        backup=Thunk(backup1_2),
        key=Thunk("backup1_2"),
        keep_meta=source1_asmt.snapshots[snap1_2.uuid].keep_meta,
    )
    return asmt


def test_print_assessment(
    golden_asmt_with_no_changes: Assessment, goldifyconsole: Console
) -> None:
    print_assessment_under_test(
        asmt=golden_asmt_with_no_changes, console=goldifyconsole
    )


def test_dont_keep(
    golden_asmt_with_no_changes: Assessment, goldifyconsole: Console
) -> None:
    # Pick an arbitrary, stable source
    _, source_asmt = sorted(golden_asmt_with_no_changes.sources.items())[0]
    # Pick an arbitrary, stable snapshot
    _, snapshot_asmt = sorted(source_asmt.snapshots.items())[0]
    # Mark the snapshot as not kept
    snapshot_asmt.keep_meta = KeepMeta()
    # Associated backup is not kept either
    backup_asmt = source_asmt.backups[snapshot_asmt.info.uuid]
    backup_asmt.keep_meta = KeepMeta()

    print_assessment_under_test(
        asmt=golden_asmt_with_no_changes, console=goldifyconsole
    )


def test_most_recent(
    golden_asmt_with_no_changes: Assessment, goldifyconsole: Console
) -> None:
    # Pick an arbitrary, stable source
    _, source_asmt = sorted(golden_asmt_with_no_changes.sources.items())[0]
    # Pick an arbitrary, stable snapshot
    _, snapshot_asmt = sorted(source_asmt.snapshots.items())[0]
    # Mark the snapshot as kept due to being the most recent
    snapshot_asmt.keep_meta = KeepMeta(reasons=Reasons.MostRecent)
    # Associated backup is kept for the same reason
    backup_asmt = source_asmt.backups[snapshot_asmt.info.uuid]
    backup_asmt.keep_meta = KeepMeta(reasons=Reasons.MostRecent)

    print_assessment_under_test(
        asmt=golden_asmt_with_no_changes, console=goldifyconsole
    )


def test_new_proposed(
    golden_asmt_with_no_changes: Assessment, goldifyconsole: Console
) -> None:
    # Pick an arbitrary, stable source
    _, source_asmt = sorted(golden_asmt_with_no_changes.sources.items())[0]
    # Create a new proposed snapshot
    proposed = mksubvol(
        id=999,
        uuid=b"\xda\x07m\x9ex\xfaLw\x89\xafYF>\xdc1\xf4",
        parent_uuid=source_asmt.info.uuid,
        flags=SubvolumeFlags.ReadOnly | SubvolumeFlags.Proposed,
        ctransid=source_asmt.info.ctransid,
        ctime=source_asmt.info.ctime,
    )
    source_asmt.snapshots[proposed.uuid] = SnapshotAssessment(
        initial_path=Path("/snapshots/proposed"),
        info=proposed,
        target_path=Thunk(lambda: Path("/snapshots/snap1_3")),  # pragma: no cover
        real_info=Thunk(lambda: proposed),  # pragma: no cover
        keep_meta=KeepMeta(reasons=Reasons.MostRecent, flags=Flags.New),
    )
    # And associated backup
    backup = backup_of_snapshot(proposed, send_parent=None)
    source_asmt.backups[backup.uuid] = BackupAssessment(
        backup=Thunk(lambda: backup),  # pragma: no cover
        key=Thunk(lambda: "backup1-3"),  # pragma: no cover
        keep_meta=source_asmt.snapshots[proposed.uuid].keep_meta,
    )

    print_assessment_under_test(
        asmt=golden_asmt_with_no_changes, console=goldifyconsole
    )


def test_backup_without_snapshot(
    golden_asmt_with_no_changes: Assessment, goldifyconsole: Console
) -> None:
    # Pick an arbitrary, stable source
    _, source_asmt = sorted(golden_asmt_with_no_changes.sources.items())[0]
    # Pick an arbitrary, stable snapshot
    uuid, _ = sorted(source_asmt.snapshots.items())[0]
    source_asmt.snapshots.pop(uuid)

    print_assessment_under_test(
        asmt=golden_asmt_with_no_changes, console=goldifyconsole
    )


def test_send_ancestor(
    golden_asmt_with_no_changes: Assessment, goldifyconsole: Console
) -> None:
    # Pick an arbitrary, stable source
    _, source_asmt = sorted(golden_asmt_with_no_changes.sources.items())[0]
    # Pick an arbitrary, stable backup which has a send-parent
    child_uuid, child_backup_asmt = sorted(
        source_asmt.backups.items(),
        key=lambda i: (i[1].backup().send_parent_uuid is None, i[0]),
    )[0]
    # Get its parent snapshot and backup
    parent_uuid = child_backup_asmt.backup().send_parent_uuid
    assert parent_uuid is not None
    parent_snapshot_asmt = source_asmt.snapshots[parent_uuid]
    parent_backup_asmt = source_asmt.backups[parent_uuid]
    # Snapshot is not kept
    parent_snapshot_asmt.keep_meta = KeepMeta()
    # Backup must be kept because it's the send-parent of another backup
    parent_backup_asmt.keep_meta = KeepMeta(
        reasons=Reasons.SendAncestor, other_uuids={child_uuid}
    )

    print_assessment_under_test(
        asmt=golden_asmt_with_no_changes, console=goldifyconsole
    )
