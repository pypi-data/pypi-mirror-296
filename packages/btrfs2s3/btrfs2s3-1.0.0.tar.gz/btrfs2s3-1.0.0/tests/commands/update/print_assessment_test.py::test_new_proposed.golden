[92mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ [0massessment and proposed new state[92m â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€[0m

[3msnapshots of /path/to/source1 in /snapshots                               [0m
 â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ 
 [1m [0m[1mpath                 [0m[1m [0m [1m [0m[1mctime              [0m[1m [0m [1m [0m[1mctransid[0m[1m [0m [1m [0m[1mpreserve     [0m[1m [0m 
 â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ 
  /snapshots/snap1_1      [34m2006[0m[34m-[0m[34m01[0m[34m-[0m[34m01[0mT[35m00[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m  [36m [0m[36m120     [0m[36m [0m  [34m2006[0m yearly    
 [2m [0m[2m/snapshots/snap1_2   [0m[2m [0m [2m [0m[2;34m2006[0m[2;34m-[0m[2;34m01[0m[2;34m-[0m[2;34m02[0m[2mT[0m[2;35m00[0m[2;35m:[0m[2;35m00[0m[2;35m:[0m[2;35m00[0m[2m [0m [2;36m [0m[2;36m122     [0m[2;36m [0m [2m [0m[2;34m2006[0m[2m yearly  [0m[2m [0m 
  âœ¨/snapshots/proposed   [34m2006[0m[34m-[0m[34m01[0m[34m-[0m[34m02[0mT[35m00[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m  [36m [0m[36m123     [0m[36m [0m  [32m<most recent>[0m  
 â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ 

tree of incremental backups of /path/to/source1 in bucket my-bucket
â”œâ”€â”€ â•­â”€ key: [32mbackup1_1[0m â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚   â”‚ [34m2006[0m[34m-[0m[34m01[0m[34m-[0m[34m01[0mT[35m00[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m / ctransid [36m120[0m                                               â”‚
â”‚   â”‚ [34m2006[0m yearly                                                                      â”‚
â”‚   â”‚ backup of: /snapshots/snap1_1                                                    â”‚
â”‚   â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
â”‚   â””â”€â”€ â•­â”€ key: [32mbackup1_2[0m â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚       â”‚ [34m2006[0m[34m-[0m[34m01[0m[34m-[0m[34m02[0mT[35m00[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m / ctransid [36m122[0m                                           â”‚
â”‚       â”‚ [34m2006[0m yearly                                                                  â”‚
â”‚       â”‚ backup of: /snapshots/snap1_2                                                â”‚
â”‚       â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
â””â”€â”€ â•­â”€ key: âœ¨[32mTBD[0m â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
    â”‚ [32m<most recent>[0m                                                                    â”‚
    â”‚ backup of: âœ¨/snapshots/proposed                                                 â”‚
    â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯


